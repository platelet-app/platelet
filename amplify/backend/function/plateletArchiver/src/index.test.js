const indexModule = require("./index");
const { default: fetch } = require("node-fetch");

const mockTaskNewer = {
    id: "anotherTaskId",
    status: "COMPLETED",
    archived: 0,
    createdAt: "2021-04-29T00:00:00.000Z",
    pickUpLocation: {
        listed: 1,
        _version: 1,
    },
    dropOffLocation: {
        listed: 0,
        _version: 1,
    },
    _version: 1,
    _deleted: null,
    _lastChangedAt: 1620000000000,
};

const mockTask = {
    id: "someTaskId",
    status: "COMPLETED",
    archived: 0,
    createdAt: "2021-04-01T00:00:00.000Z",
    pickUpLocation: {
        listed: 1,
        _version: 1,
        id: "someLocationIdListed",
    },
    dropOffLocation: {
        id: "someLocationIdUnlisted",
        listed: 0,
        _version: 1,
    },
    _version: 1,
    _deleted: null,
    _lastChangedAt: 1620000000000,
};

const fakeData = {
    tasksByArchivedStatus: {
        items: [mockTask, mockTaskNewer],
        nextToken: "tasktoken",
    },
};
const fakeDataSecond = {
    tasksByArchivedStatus: {
        items: [mockTaskNewer],
        nextToken: null,
    },
};
const fakeEmptyData = {
    tasksByArchivedStatus: {
        items: [],
        nextToken: null,
    },
};
const fakeAssigneeData = {
    getTask: {
        assignees: {
            items: [
                {
                    id: "someAssignmentId",
                    _version: 1,
                },
            ],
            nextToken: "asstoken",
        },
    },
};
const fakeAssigneeDataSecond = {
    ...fakeAssigneeData,
    getTask: {
        ...fakeAssigneeData.getTask,
        assignees: {
            ...fakeAssigneeData.getTask.assignees,
            nextToken: null,
        },
    },
};
const fakeAssigneeReturn = {
    updateTaskAssignee: {
        id: "someAssignmentId",
        archived: 1,
    },
};
const fakeDeliverableData = {
    getTask: {
        deliverables: {
            items: [
                {
                    id: "someDeliverableId",
                    _version: 1,
                },
            ],
            nextToken: "deltoken",
        },
    },
};
const fakeDeliverableDataSecond = {
    ...fakeDeliverableData,
    getTask: {
        ...fakeDeliverableData.getTask,
        deliverables: {
            ...fakeDeliverableData.getTask.deliverables,
            nextToken: null,
        },
    },
};
const fakeDeliverableReturn = {
    updateDeliverable: {
        id: "someDeliverableId",
        archived: 1,
    },
};
const fakeCommentData = {
    getTask: {
        comments: {
            items: [
                {
                    id: "someCommentId",
                    _version: 1,
                },
            ],
            nextToken: "comtoken",
        },
    },
};
const fakeCommentDataSecond = {
    ...fakeCommentData,
    getTask: {
        ...fakeCommentData.getTask,
        comments: {
            ...fakeCommentData.getTask.comments,
            nextToken: null,
        },
    },
};
const fakeCommentReturn = {
    updateComment: {
        id: "someCommentId",
        archived: 1,
    },
};
const fakeTaskReturn = {
    updateTask: {
        id: "someTaskId",
        archived: 1,
    },
};

function setupFetchStub(data) {
    return function fetchStub(_url) {
        return new Promise((resolve) => {
            resolve({
                json: () =>
                    Promise.resolve({
                        data,
                    }),
            });
        });
    };
}

describe("plateletArchiver", () => {
    afterEach(() => jest.restoreAllMocks());

    beforeAll(() => {
        jest.useFakeTimers("modern");
        jest.setSystemTime(new Date("2021-05-01T00:00:00.000Z"));
    });

    it("should return a function", () => {
        expect(typeof indexModule.handler).toBe("function");
    });

    test("archive some tasks, ignore the newer task", async () => {
        const requestSpy = jest.spyOn(indexModule, "makeNewRequest");
        jest.spyOn(fetch, "default")
            .mockImplementationOnce(setupFetchStub(fakeData))
            .mockImplementationOnce(setupFetchStub(fakeDataSecond))
            .mockImplementationOnce(setupFetchStub(fakeEmptyData))
            .mockImplementationOnce(setupFetchStub(fakeEmptyData))
            .mockImplementationOnce(setupFetchStub(fakeEmptyData))
            .mockImplementationOnce(setupFetchStub({}))
            .mockImplementationOnce(setupFetchStub(fakeAssigneeData))
            .mockImplementationOnce(setupFetchStub(fakeAssigneeDataSecond))
            .mockImplementationOnce(setupFetchStub(fakeAssigneeReturn))
            .mockImplementationOnce(setupFetchStub(fakeAssigneeReturn))
            .mockImplementationOnce(setupFetchStub(fakeDeliverableData))
            .mockImplementationOnce(setupFetchStub(fakeDeliverableDataSecond))
            .mockImplementationOnce(setupFetchStub(fakeDeliverableReturn))
            .mockImplementationOnce(setupFetchStub(fakeDeliverableReturn))
            .mockImplementationOnce(setupFetchStub(fakeCommentData))
            .mockImplementationOnce(setupFetchStub(fakeCommentDataSecond))
            .mockImplementationOnce(setupFetchStub(fakeCommentReturn))
            .mockImplementationOnce(setupFetchStub(fakeCommentReturn))
            .mockImplementationOnce(setupFetchStub(fakeTaskReturn));
        await indexModule.handler({}, indexModule.makeNewRequest);
        expect(requestSpy).toMatchSnapshot();
        expect(requestSpy).not.toHaveBeenCalledWith(
            expect.anything(),
            expect.objectContaining({ id: "anotherTaskId" })
        );
    });
    test("failure to archive an assignee", async () => {
        const unArchivedFakeAssigneeReturn = {
            updateTaskAssignee: {
                id: "someAssignmentId",
                archived: 0,
            },
        };
        const requestSpy = jest.spyOn(indexModule, "makeNewRequest");
        jest.spyOn(fetch, "default")
            .mockImplementationOnce(setupFetchStub(fakeData))
            .mockImplementationOnce(setupFetchStub(fakeDataSecond))
            .mockImplementationOnce(setupFetchStub(fakeEmptyData))
            .mockImplementationOnce(setupFetchStub(fakeEmptyData))
            .mockImplementationOnce(setupFetchStub(fakeEmptyData))
            .mockImplementationOnce(setupFetchStub({}))
            .mockImplementationOnce(setupFetchStub(fakeAssigneeDataSecond))
            .mockImplementationOnce(
                setupFetchStub(unArchivedFakeAssigneeReturn)
            );
        await indexModule.handler({}, indexModule.makeNewRequest);
        expect(requestSpy).toMatchSnapshot();
        expect(requestSpy).not.toHaveBeenCalledWith(
            expect.stringMatching(/updateTask\(/),
            expect.objectContaining({ id: "someTaskId" })
        );
    });
});
