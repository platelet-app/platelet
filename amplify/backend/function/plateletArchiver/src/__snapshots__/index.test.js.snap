// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plateletArchiver archive some tasks, ignore the newer task 1`] = `
[MockFunction] {
  "calls": [
    [
      "
    query LIST_TASKS_BY_ARCHIVE_STATUS(
        $archived: Int!
        $status: ModelStringKeyConditionInput
        $nextToken: String
    ) {
        tasksByArchivedStatus(
            archived: $archived
            status: $status
            nextToken: $nextToken
        ) {
            items {
                id
                createdAt
                establishmentLocation {
                    id
                    listed
                    _version
                }
                pickUpLocation {
                    id
                    listed
                    _version
                }
                dropOffLocation {
                    id
                    listed
                    _version
                }
                _version
            }
            nextToken
        }
    }
",
      {
        "archived": 0,
        "nextToken": null,
        "status": {
          "eq": "COMPLETED",
        },
      },
    ],
    [
      "
    query LIST_TASKS_BY_ARCHIVE_STATUS(
        $archived: Int!
        $status: ModelStringKeyConditionInput
        $nextToken: String
    ) {
        tasksByArchivedStatus(
            archived: $archived
            status: $status
            nextToken: $nextToken
        ) {
            items {
                id
                createdAt
                establishmentLocation {
                    id
                    listed
                    _version
                }
                pickUpLocation {
                    id
                    listed
                    _version
                }
                dropOffLocation {
                    id
                    listed
                    _version
                }
                _version
            }
            nextToken
        }
    }
",
      {
        "archived": 0,
        "nextToken": null,
        "status": {
          "eq": "REJECTED",
        },
      },
    ],
    [
      "
    query LIST_TASKS_BY_ARCHIVE_STATUS(
        $archived: Int!
        $status: ModelStringKeyConditionInput
        $nextToken: String
    ) {
        tasksByArchivedStatus(
            archived: $archived
            status: $status
            nextToken: $nextToken
        ) {
            items {
                id
                createdAt
                establishmentLocation {
                    id
                    listed
                    _version
                }
                pickUpLocation {
                    id
                    listed
                    _version
                }
                dropOffLocation {
                    id
                    listed
                    _version
                }
                _version
            }
            nextToken
        }
    }
",
      {
        "archived": 0,
        "nextToken": null,
        "status": {
          "eq": "ABANDONED",
        },
      },
    ],
    [
      "
    query LIST_TASKS_BY_ARCHIVE_STATUS(
        $archived: Int!
        $status: ModelStringKeyConditionInput
        $nextToken: String
    ) {
        tasksByArchivedStatus(
            archived: $archived
            status: $status
            nextToken: $nextToken
        ) {
            items {
                id
                createdAt
                establishmentLocation {
                    id
                    listed
                    _version
                }
                pickUpLocation {
                    id
                    listed
                    _version
                }
                dropOffLocation {
                    id
                    listed
                    _version
                }
                _version
            }
            nextToken
        }
    }
",
      {
        "archived": 0,
        "nextToken": null,
        "status": {
          "eq": "CANCELLED",
        },
      },
    ],
    [
      "
    query LIST_TASKS_BY_ARCHIVE_STATUS(
        $archived: Int!
        $status: ModelStringKeyConditionInput
        $nextToken: String
    ) {
        tasksByArchivedStatus(
            archived: $archived
            status: $status
            nextToken: $nextToken
        ) {
            items {
                id
                createdAt
                establishmentLocation {
                    id
                    listed
                    _version
                }
                pickUpLocation {
                    id
                    listed
                    _version
                }
                dropOffLocation {
                    id
                    listed
                    _version
                }
                _version
            }
            nextToken
        }
    }
",
      {
        "archived": 0,
        "nextToken": "tasktoken",
        "status": {
          "eq": "COMPLETED",
        },
      },
    ],
    [
      "
    mutation UpdateLocation($input: UpdateLocationInput!) {
        updateLocation(input: $input) {
            id
            archived
        };
    };
",
      {
        "input": {
          "_version": 1,
          "archived": 1,
          "id": "someLocationIdUnlisted",
        },
      },
    ],
    [
      "
    query GetTaskAssignees(
        $id: ID!,
        $nextToken: String
    ) {
        getTask(id: $id) {
            id
            assignees(nextToken: $nextToken) {
                items {
                    id
                    _version
                }
                nextToken
            };
        };
    };
",
      {
        "id": "someTaskId",
        "nextToken": null,
      },
    ],
    [
      "
    query GetTaskAssignees(
        $id: ID!,
        $nextToken: String
    ) {
        getTask(id: $id) {
            id
            assignees(nextToken: $nextToken) {
                items {
                    id
                    _version
                }
                nextToken
            };
        };
    };
",
      {
        "id": "someTaskId",
        "nextToken": "asstoken",
      },
    ],
    [
      "
    mutation UpdateTaskAssignee($input: UpdateTaskAssigneeInput!) {
        updateTaskAssignee(input: $input) {
            id
            archived
        };
    };
",
      {
        "input": {
          "_version": 1,
          "archived": 1,
          "id": "someAssignmentId",
        },
      },
    ],
    [
      "
    mutation UpdateTaskAssignee($input: UpdateTaskAssigneeInput!) {
        updateTaskAssignee(input: $input) {
            id
            archived
        };
    };
",
      {
        "input": {
          "_version": 1,
          "archived": 1,
          "id": "someAssignmentId",
        },
      },
    ],
    [
      "
    query GetTaskDeliverables($id: ID!, $nextToken: String) {
        getTask(id: $id) {
            id
            deliverables(nextToken: $nextToken) {
                items {
                    id
                    _version
                }
                nextToken
            };
        };
    };
",
      {
        "id": "someTaskId",
        "nextToken": null,
      },
    ],
    [
      "
    query GetTaskDeliverables($id: ID!, $nextToken: String) {
        getTask(id: $id) {
            id
            deliverables(nextToken: $nextToken) {
                items {
                    id
                    _version
                }
                nextToken
            };
        };
    };
",
      {
        "id": "someTaskId",
        "nextToken": "deltoken",
      },
    ],
    [
      "
    mutation UpdateDeliverable($input: UpdateDeliverableInput!) {
        updateDeliverable(input: $input) {
            id
            archived
        };
    };
",
      {
        "input": {
          "_version": 1,
          "archived": 1,
          "id": "someDeliverableId",
        },
      },
    ],
    [
      "
    mutation UpdateDeliverable($input: UpdateDeliverableInput!) {
        updateDeliverable(input: $input) {
            id
            archived
        };
    };
",
      {
        "input": {
          "_version": 1,
          "archived": 1,
          "id": "someDeliverableId",
        },
      },
    ],
    [
      "
    query GetTaskComments($id: ID!, $nextToken: String) {
        getTask(id: $id) {
            id
            comments(nextToken: $nextToken) {
                items {
                    id
                    _version
                }
                nextToken
            };
        };
    };
",
      {
        "id": "someTaskId",
        "nextToken": null,
      },
    ],
    [
      "
    query GetTaskComments($id: ID!, $nextToken: String) {
        getTask(id: $id) {
            id
            comments(nextToken: $nextToken) {
                items {
                    id
                    _version
                }
                nextToken
            };
        };
    };
",
      {
        "id": "someTaskId",
        "nextToken": "comtoken",
      },
    ],
    [
      "
    mutation UpdateComment($input: UpdateCommentInput!) {
        updateComment(input: $input) {
            id
            archived
        };
    };
",
      {
        "input": {
          "_version": 1,
          "archived": 1,
          "id": "someCommentId",
        },
      },
    ],
    [
      "
    mutation UpdateComment($input: UpdateCommentInput!) {
        updateComment(input: $input) {
            id
            archived
        };
    };
",
      {
        "input": {
          "_version": 1,
          "archived": 1,
          "id": "someCommentId",
        },
      },
    ],
    [
      "
    mutation UpdateTask($input: UpdateTaskInput!) {
        updateTask(input: $input) {
            id
            archived
        }
    }
",
      {
        "input": {
          "_version": 1,
          "archived": 1,
          "id": "someTaskId",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`plateletArchiver failure to archive an assignee 1`] = `
[MockFunction] {
  "calls": [
    [
      "
    query LIST_TASKS_BY_ARCHIVE_STATUS(
        $archived: Int!
        $status: ModelStringKeyConditionInput
        $nextToken: String
    ) {
        tasksByArchivedStatus(
            archived: $archived
            status: $status
            nextToken: $nextToken
        ) {
            items {
                id
                createdAt
                establishmentLocation {
                    id
                    listed
                    _version
                }
                pickUpLocation {
                    id
                    listed
                    _version
                }
                dropOffLocation {
                    id
                    listed
                    _version
                }
                _version
            }
            nextToken
        }
    }
",
      {
        "archived": 0,
        "nextToken": null,
        "status": {
          "eq": "COMPLETED",
        },
      },
    ],
    [
      "
    query LIST_TASKS_BY_ARCHIVE_STATUS(
        $archived: Int!
        $status: ModelStringKeyConditionInput
        $nextToken: String
    ) {
        tasksByArchivedStatus(
            archived: $archived
            status: $status
            nextToken: $nextToken
        ) {
            items {
                id
                createdAt
                establishmentLocation {
                    id
                    listed
                    _version
                }
                pickUpLocation {
                    id
                    listed
                    _version
                }
                dropOffLocation {
                    id
                    listed
                    _version
                }
                _version
            }
            nextToken
        }
    }
",
      {
        "archived": 0,
        "nextToken": null,
        "status": {
          "eq": "REJECTED",
        },
      },
    ],
    [
      "
    query LIST_TASKS_BY_ARCHIVE_STATUS(
        $archived: Int!
        $status: ModelStringKeyConditionInput
        $nextToken: String
    ) {
        tasksByArchivedStatus(
            archived: $archived
            status: $status
            nextToken: $nextToken
        ) {
            items {
                id
                createdAt
                establishmentLocation {
                    id
                    listed
                    _version
                }
                pickUpLocation {
                    id
                    listed
                    _version
                }
                dropOffLocation {
                    id
                    listed
                    _version
                }
                _version
            }
            nextToken
        }
    }
",
      {
        "archived": 0,
        "nextToken": null,
        "status": {
          "eq": "ABANDONED",
        },
      },
    ],
    [
      "
    query LIST_TASKS_BY_ARCHIVE_STATUS(
        $archived: Int!
        $status: ModelStringKeyConditionInput
        $nextToken: String
    ) {
        tasksByArchivedStatus(
            archived: $archived
            status: $status
            nextToken: $nextToken
        ) {
            items {
                id
                createdAt
                establishmentLocation {
                    id
                    listed
                    _version
                }
                pickUpLocation {
                    id
                    listed
                    _version
                }
                dropOffLocation {
                    id
                    listed
                    _version
                }
                _version
            }
            nextToken
        }
    }
",
      {
        "archived": 0,
        "nextToken": null,
        "status": {
          "eq": "CANCELLED",
        },
      },
    ],
    [
      "
    query LIST_TASKS_BY_ARCHIVE_STATUS(
        $archived: Int!
        $status: ModelStringKeyConditionInput
        $nextToken: String
    ) {
        tasksByArchivedStatus(
            archived: $archived
            status: $status
            nextToken: $nextToken
        ) {
            items {
                id
                createdAt
                establishmentLocation {
                    id
                    listed
                    _version
                }
                pickUpLocation {
                    id
                    listed
                    _version
                }
                dropOffLocation {
                    id
                    listed
                    _version
                }
                _version
            }
            nextToken
        }
    }
",
      {
        "archived": 0,
        "nextToken": "tasktoken",
        "status": {
          "eq": "COMPLETED",
        },
      },
    ],
    [
      "
    mutation UpdateLocation($input: UpdateLocationInput!) {
        updateLocation(input: $input) {
            id
            archived
        };
    };
",
      {
        "input": {
          "_version": 1,
          "archived": 1,
          "id": "someLocationIdUnlisted",
        },
      },
    ],
    [
      "
    query GetTaskAssignees(
        $id: ID!,
        $nextToken: String
    ) {
        getTask(id: $id) {
            id
            assignees(nextToken: $nextToken) {
                items {
                    id
                    _version
                }
                nextToken
            };
        };
    };
",
      {
        "id": "someTaskId",
        "nextToken": null,
      },
    ],
    [
      "
    mutation UpdateTaskAssignee($input: UpdateTaskAssigneeInput!) {
        updateTaskAssignee(input: $input) {
            id
            archived
        };
    };
",
      {
        "input": {
          "_version": 1,
          "archived": 1,
          "id": "someAssignmentId",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
